### PR TITLE
[REVIEW] Fix race condition in device_scalar::set_value

### PR DESCRIPTION
**Pull request by jrhemstad**
`device_scalar::set_value` performs a host to device copy to set the value of the contained object. This copy is performed asynchronously and does not perform any synchronization.

Previously, the `host_value` was being passed by value. This is problematic when combined with the fact that the copy was asynchronous. This could mean that the copy would not complete before the function local copy of `host_value` was destroyed. 

I remedied this by making it pass by reference and clarifying the documentation about the asynchronous behavior.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Review by rongou — APPROVED** [review]

File: include/rmm/device_scalar.hpp [hunk 1]
```diff
@@ -114,7 +114,25 @@ class device_scalar {
    * (e.g. using `cudaStreamWaitEvent()` or `cudaStreamSynchronize()`) before and after calling
    * this function, otherwise there may be a race condition.
    *
-   * Does not synchronize `stream`.
+   * This function does not synchronize stream `stream` before returning. Therefore, the object
```
Comment (harrism) [inline include/rmm/device_scalar.hpp:117]: Delete first occurrence of the word stream (on mobile, suggestions don't work)

File: include/rmm/device_scalar.hpp [hunk 2]
```diff
@@ -141,7 +159,25 @@ class device_scalar {
...
```
Comment (harrism) [inline include/rmm/device_scalar.hpp:162]: Also here.

**Review by harrism — CHANGES_REQUESTED** [review]
Should we add _async to the name of set_value() in this PR?

File: include/rmm/device_scalar.hpp [hunk 3]
```diff
@@ -123,7 +141,7 @@ class device_scalar {
    * @param stream CUDA stream on which to perform the copy
    */
   template <typename Dummy = void>
-  auto set_value(T host_value, cudaStream_t stream = 0)
+  auto set_value(T const &host_value, cudaStream_t stream = 0)
     -> std::enable_if_t<std::is_fundamental<T>::value, Dummy>
```
Comment (cwharris) [inline include/rmm/device_scalar.hpp:145]: Let's rename `Dummy` => `Placeholder` or something, while we're here.

File: include/rmm/device_scalar.hpp [hunk 4]
```diff
@@ -150,7 +186,7 @@ class device_scalar {
    * @param stream CUDA stream on which to perform the copy
    */
   template <typename Dummy = void>
-  auto set_value(T host_value, cudaStream_t stream = 0)
+  auto set_value(T const &host_value, cudaStream_t stream = 0)
     -> std::enable_if_t<not std::is_fundamental<T>::value, Dummy>
```
Comment (cwharris) [inline include/rmm/device_scalar.hpp:190]: same here.

**Review by cwharris — CHANGES_REQUESTED** [review]

**Comment by jrhemstad** [comment]
> Should we add _async to the name of set_value() in this PR?

That would break code in cuDF and elsewhere and I didn't want to deal with that in this PR. See https://github.com/rapidsai/rmm/issues/570

File: include/rmm/device_scalar.hpp (hunk 3 above)
Comment (harrism) [reply include/rmm/device_scalar.hpp:145]: Yes, good catch.

**Review by harrism — APPROVED** [review]

**Comment by jrhemstad** [comment]
rerun tests

**Comment by jrhemstad** [comment]
@cwharris need approval

**Review by cwharris — APPROVED** [review]
lgtm

**Pull request merged**